modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/MoveKBObjectMVCActionCommand.java

```diff
@@ -52,31 +64,113 @@ protected void doProcessAction(
-			KBArticleConstants.getClassName());
+		boolean dragAndDrop = ParamUtil.getBoolean(
+			actionRequest, "dragAndDrop");
 
-		if (resourceClassNameId == kbArticleClassNameId) {
+		try {
+			long resourceClassNameId = ParamUtil.getLong(
+				actionRequest, "resourceClassNameId");
+			long resourcePrimKey = ParamUtil.getLong(
+				actionRequest, "resourcePrimKey");
 			long parentResourceClassNameId = ParamUtil.getLong(
 				actionRequest, "parentResourceClassNameId",
 				_portal.getClassNameId(KBFolderConstants.getClassName()));
-			double priority = ParamUtil.getDouble(actionRequest, "priority");
+			long parentResourcePrimKey = ParamUtil.getLong(
+				actionRequest, "parentResourcePrimKey",
+				KBFolderConstants.DEFAULT_PARENT_FOLDER_ID);
+
+			long kbArticleClassNameId = _portal.getClassNameId(
+				KBArticleConstants.getClassName());
+
+			if (resourceClassNameId == kbArticleClassNameId) {
+				if (!_isDragAndDrop(dragAndDrop)) {
+					double priority = ParamUtil.getDouble(
+						actionRequest, "priority");
+
+					_kbArticleService.moveKBArticle(
+						resourcePrimKey, parentResourceClassNameId,
+						parentResourcePrimKey, priority);
+				}
+				else {
+					KBArticle kbArticle = _kbArticleService.getLatestKBArticle(
+						resourcePrimKey, WorkflowConstants.STATUS_ANY);
+
+					if (kbArticle.getParentResourcePrimKey() !=
+							parentResourcePrimKey) {
+
+						_kbArticleService.moveKBArticle(
+							resourcePrimKey, parentResourceClassNameId,
+							parentResourcePrimKey, kbArticle.getPriority());
+					}
+				}
+			}
+			else {
+				if (!_isDragAndDrop(dragAndDrop)) {
+					_kbFolderService.moveKBFolder(
+						resourcePrimKey, parentResourcePrimKey);
+				}
+				else {
+					if (parentResourceClassNameId == kbArticleClassNameId) {
+						_errorMessage(
+							actionRequest, actionResponse,
+							_language.get(
+								_portal.getHttpServletRequest(actionRequest),
+								"folders-can-not-be-moved-into-articles"));
 
-			_kbArticleService.moveKBArticle(
-				resourcePrimKey, parentResourceClassNameId,
-				parentResourcePrimKey, priority);
+						return;
+					}
+
+					KBFolder kbFolder = _kbFolderService.getKBFolder(
+						resourcePrimKey);
+
+					if (kbFolder.getParentKBFolderId() !=
+							parentResourcePrimKey) {
+
+						_kbFolderService.moveKBFolder(
+							resourcePrimKey, parentResourcePrimKey);
+					}
+				}
+			}
 		}
-		else {
-			_kbFolderService.moveKBFolder(
-				resourcePrimKey, parentResourcePrimKey);
+		catch (PortalException portalException) {
+			if (!_isDragAndDrop(dragAndDrop)) {
+				throw portalException;
+			}
+
+			_errorMessage(
+				actionRequest, actionResponse,
+				_language.get(
+					_portal.getHttpServletRequest(actionRequest),
+					"your-request-failed-to-complete"));
+
+			return;
 		}
+
+		if (_isDragAndDrop(dragAndDrop)) {
+			JSONObject jsonObject = JSONUtil.put("success", Boolean.TRUE);
+
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse, jsonObject);
+		}
```

Comment:
Instead of putting this after the try/catch, why not move it inside the try/catch block (as the last action)? That would allow you to remove the `return` statement from the `catch` block.